src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs

```diff
@@ -76,15 +79,24 @@ public async Task StartAsync(CancellationToken cancellationToken = default)
             _logger.Started();
         }
 
-        private async Task HandleBackgroundException(BackgroundService backgroundService)
+        private async Task TryStartBackgroundServiceAsync(
+            BackgroundService backgroundService, CancellationToken cancellationToken)
         {
             try
             {
+                // Fire IHostedService.Start
+                await backgroundService.StartAsync(cancellationToken).ConfigureAwait(false);
                 await backgroundService.ExecuteTask.ConfigureAwait(false);
             }
             catch (Exception ex)
             {
                 _logger.BackgroundServiceFaulted(ex);
+                if (_options is { BackgroundServiceExceptionBehavior: BackgroundServiceExceptionBehavior.StopHost })
+                {
+                    _logger.LogCritical(SR.Format(SR.BackgroundServiceExceptionStoppedHost, ex));
+
+                    await StopAsync(CancellationToken.None).ConfigureAwait(false);
```

Comment:
This shouldn't trigger stop, we should run through the normal code path. This should trigger `IHostApplicationLifetime.StopApplication`.

Reply:
But this call to `StopAsync` internally calls `IHostApplicationLifetime.StopApplication`, it additionally performs explicit clean up.

Reply:
I'm not sure we should do this here. I feel like we should stop the host and let it react the way it typically does.

Reply:
Then `IHostApplicationLifetime.StopApplication` will be called twice, is that ok?

Reply:
StopApplication triggers the host unwind which then calls StopAsync on the host. This only works when you call host.Run (which is fine). Yes StopApplication would get called twice but it'll only fire once.

Reply:
When a user has `HostOptions.BackgroundServiceExceptionBehavior` set to `BackgroundServiceExceptionBehavior.StopHost` and a `BackgroundService` throws an unhandled exception, do you expect that the `Host` to trigger the `IApplicationLifetime.ApplicationStarted` event?

Calling `StopApplication`, while it does trigger the `IApplicationLifetime.ApplicationStopping` it doesn't actually stop the host. The host continues, and we end up with the host firing `ApplicationStopping`, then `ApplicationStarting` in that order. And it's not actually stopped.

> This only works when you call host.Run (which is fine)

Wouldn't we want consistent behavior, regardless of how the host is started? If I'm understanding this correctly, calling `host.Start` would work differently than calling `host.Run` with regard to how this new `BackgroundServiceExceptionBehavior` actually behaves - which is alarming to me. What am I missing?

Reply:
The bigger issue you raise is how StopApplication works. It needs to work regardless of how the host was started. Today it relies on using something like Run/RunAsync to trigger the call to StopAsync on the host. That said, I think it's a separate issue. Stopping the host should be triggered the same way that any piece of code can do by using IHostApplicationLifetime.StopApplication.

Reply:
Feedback addressed in https://github.com/dotnet/runtime/pull/50569/commits/e985ec5dbf26f269c33c39fea3b016764280077b